src/core/lib/gprpp/ref_counted.h

```diff
@@ -34,6 +36,63 @@
 
 namespace grpc_core {
 
+// RefCount is a simple atomic ref-count.
+//
+// This is a C++ implementation of gpr_refcount, with inline functions. Due to
+// inline functions, this class is significantly more efficient than
+// gpr_refcount and hence is a major performance win.
+//
+// TODO(soheil): Remove gpr_refcount after submitting the GRFC and the paragraph
+//               above.
+class RefCount {
+ public:
+  // `init` is the initial refcount stored in this object.
+  explicit RefCount(intptr_t init = 1) : val_(init) {}
+
+  // Reinitializes the ref-count.
+  void Reset(intptr_t init) { val_.store(init, std::memory_order_relaxed); }
+
+  // Increases the ref-count by `n`.
+  void Ref(intptr_t n = 1) { val_.fetch_add(n, std::memory_order_relaxed); }
+
+  // Similar to Ref() with an assert on the ref-count being non-zero.
+  void RefNonZero() {
+#ifndef NDEBUG
+    const intptr_t prior = val_.fetch_add(1, std::memory_order_relaxed);
+    assert(prior > 0);
+#else
+    Ref();
+#endif
+  }
+
+  // Increments the ref-count only if the ref-count is not 0.
+  bool RefIfNonZero() {
+    intptr_t prior = val_.load(std::memory_order_relaxed);
+    do {
+      if (prior == 0) return false;
+    } while (!val_.compare_exchange_weak(prior, prior + 1,
+                                        std::memory_order_release,
+                                        std::memory_order_relaxed));
+    return true;
+  }
+
+  // Returns whether there is only one reference tracked in this refcount.
+  bool IsUnique() const { return val_.load(std::memory_order_acquire); }
+
+  // Decrements the ref-count and returns true if the ref-count reaches 0.
+  bool Unref() {
+    const intptr_t prior = val_.fetch_sub(1, std::memory_order_acq_rel);
+    GPR_DEBUG_ASSERT(prior > 0);
+    return prior == 1;
+  }
+
+  // Returns the current value of the ref-count without any ordering guarantees.
+  intptr_t get() const { return val_.load(std::memory_order_relaxed); }
```

Comment:
This probably shouldn't be a public method.  It looks like the main use-case is `RefCountedWithTracing` below.  How about simply making that class a friend of this?

Might need to do the same thing for `InternallyRefCountedWithTracing`, which is defined in orphanable.h.